File: frontend/src/schemas/clinic.ts
/**
 * Zod schemas — Clinic module.
 */
import { z } from 'zod';

export const TRIAGE_PRIORITIES = ['low', 'medium', 'high', 'urgent'] as const;
export type TriagePriority = (typeof TRIAGE_PRIORITIES)[number];

export const encounterSchema = z.object({
  id: z.number().int().positive(),
  patient_school_id: z.string(),
  chief_complaint: z.string(),
  triage_priority: z.enum(TRIAGE_PRIORITIES).nullable().optional(),
  triage_override: z.boolean().optional(),
  diagnosis: z.string().nullable().optional(),
  status: z.enum(['Open', 'Closed', 'Referred']),
  attending_user_id: z.number().int().positive(),
  started_at: z.string(),
  closed_at: z.string().nullable(),
});
export type Encounter = z.infer<typeof encounterSchema>;

export const TREATMENT_TYPES = ['medication', 'first_aid', 'procedure', 'referral', 'other'] as const;
export type TreatmentType = (typeof TREATMENT_TYPES)[number];

export const treatmentSchema = z.object({
  id: z.number().int().positive(),
  encounter_id: z.number().int().positive(),
  treatment_type: z.enum(TREATMENT_TYPES),
  description: z.string(),
  medicine_id: z.number().int().nullable(),
  medicine_name: z.string().nullable().optional(),
  unit: z.string().nullable().optional(),
  quantity_used: z.number().int().nullable(),
  administered_at: z.string(),
});
export type Treatment = z.infer<typeof treatmentSchema>;

export const setAssessmentSchema = z.object({
  triage_priority: z.enum(TRIAGE_PRIORITIES).optional(),
  diagnosis: z.string().max(5000).optional().or(z.literal('')),
});
export type SetAssessmentInput = z.infer<typeof setAssessmentSchema>;

export const addTreatmentSchema = z
  .object({
    treatment_type: z.enum(TREATMENT_TYPES),
    description: z.string().min(1, 'Required.').max(2000),
    medicine_id: z.number().int().positive().optional(),
    quantity: z.number().int().positive().optional(),
  })
  .refine((v) => v.treatment_type !== 'medication' || (v.medicine_id !== undefined && v.quantity !== undefined), {
    message: 'Medicine and quantity are required for a medication treatment.',
    path: ['medicine_id'],
  });
export type AddTreatmentInput = z.infer<typeof addTreatmentSchema>;

export const triagePredictionSchema = z.object({
  id: z.number().int().positive(),
  encounter_id: z.number().int().positive(),
  predicted_priority: z.enum(TRIAGE_PRIORITIES),
  confidence_score: z.number(),
  model_version: z.string(),
  features_used: z.record(z.string(), z.unknown()),
  staff_decision: z.enum(['accepted', 'overridden']).nullable(),
});
export type TriagePrediction = z.infer<typeof triagePredictionSchema>;

export const createEncounterSchema = z.object({
  patient_school_id: z.string().min(1).max(32),
  chief_complaint: z.string().min(1).max(255),
});
export type CreateEncounterInput = z.infer<typeof createEncounterSchema>;

export const vitalsSchema = z.object({
  encounter_id: z.number().int().positive(),
  bp_systolic: z.number().int().nullable(),
  bp_diastolic: z.number().int().nullable(),
  pulse_bpm: z.number().int().nullable(),
  temp_c: z.number().nullable(),
  spo2_pct: z.number().int().nullable(),
  weight_kg: z.number().nullable(),
  height_cm: z.number().nullable(),
  recorded_at: z.string(),
});
export type Vitals = z.infer<typeof vitalsSchema>;

export const recordVitalsSchema = z.object({
  bp_systolic:  z.number().int().min(0).max(300).optional(),
  bp_diastolic: z.number().int().min(0).max(200).optional(),
  pulse_bpm:    z.number().int().min(0).max(300).optional(),
  temp_c:       z.number().min(20).max(45).optional(),
  spo2_pct:     z.number().int().min(0).max(100).optional(),
  weight_kg:    z.number().min(0).max(600).optional(),
  height_cm:    z.number().min(0).max(300).optional(),
});
export type RecordVitalsInput = z.infer<typeof recordVitalsSchema>;

/**
 * Result of a bulk CSV import. Mirrors `bulkImportEncounters` return
 * shape on the backend (`Modules\Clinic\Services\ClinicService`).
 */
export const importEncountersResultSchema = z.object({
  imported: z.number().int().min(0),
  first_id: z.number().int().positive(),
  last_id: z.number().int().positive(),
});
export type ImportEncountersResult = z.infer<typeof importEncountersResultSchema>;

File: mobile/lib/core/models/clinic.dart
import 'referral.dart';

class ClinicEncounter {
  ClinicEncounter({
    required this.id,
    required this.patientSchoolId,
    required this.chiefComplaint,
    required this.status,
    required this.attendingUserId,
    required this.startedAt,
    this.patientName,
    this.appointmentId,
    this.stationId,
    this.triagePriority,
    this.triageOverride = false,
    this.diagnosis,
    this.outcome,
    this.closedAt,
    this.queueEntryId,
    this.queueNumber,
    this.queueStatus,
    this.incomingReferralId,
    this.vitalsCount = 0,
    this.treatmentCount = 0,
    this.assessmentRecorded = false,
    this.outgoingReferral,
  });

  factory ClinicEncounter.fromJson(Map<String, dynamic> json) =>
      ClinicEncounter(
        id: (json['id'] ?? 0) as int,
        patientSchoolId: (json['patient_school_id'] ?? '') as String,
        patientName: json['patient_name'] as String?,
        appointmentId: json['appointment_id'] as int?,
        stationId: json['station_id'] as String?,
        chiefComplaint: (json['chief_complaint'] ?? '') as String,
        triagePriority: json['triage_priority'] as String?,
        triageOverride: (json['triage_override'] ?? false) as bool,
        diagnosis: json['diagnosis'] as String?,
        status: (json['status'] ?? '') as String,
        outcome: json['outcome'] as String?,
        attendingUserId: (json['attending_user_id'] ?? 0) as int,
        startedAt: (json['started_at'] ?? '') as String,
        closedAt: json['closed_at'] as String?,
        queueEntryId: json['queue_entry_id'] as int?,
        queueNumber: json['queue_number'] as String?,
        queueStatus: json['queue_status'] as String?,
        incomingReferralId: json['incoming_referral_id'] as int?,
        vitalsCount: (json['vitals_count'] ?? 0) as int,
        treatmentCount: (json['treatment_count'] ?? 0) as int,
        assessmentRecorded: (json['assessment_recorded'] ?? false) as bool,
        outgoingReferral: json['outgoing_referral'] is Map<String, dynamic>
            ? Referral.fromJson(
                json['outgoing_referral'] as Map<String, dynamic>)
            : null,
      );

  final int id;
  final String patientSchoolId;
  final String? patientName;
  final int? appointmentId;
  final String? stationId;
  final String chiefComplaint;
  final String? triagePriority;
  final bool triageOverride;
  final String? diagnosis;
  final String status;
  final String? outcome;
  final int attendingUserId;
  final String startedAt;
  final String? closedAt;
  final int? queueEntryId;
  final String? queueNumber;
  final String? queueStatus;
  final int? incomingReferralId;
  final int vitalsCount;
  final int treatmentCount;
  final bool assessmentRecorded;
  final Referral? outgoingReferral;

  bool get isOpen => status == 'open';
  String get patientDisplayName =>
      patientName?.trim().isNotEmpty == true ? patientName! : patientSchoolId;
}

class ClinicVitals {
  ClinicVitals({
    required this.encounterId,
    required this.recordedAt,
    this.bpSystolic,
    this.bpDiastolic,
    this.pulseBpm,
    this.tempC,
    this.spo2Pct,
    this.weightKg,
    this.heightCm,
  });

  factory ClinicVitals.fromJson(Map<String, dynamic> json) => ClinicVitals(
        encounterId: (json['encounter_id'] ?? 0) as int,
        bpSystolic: (json['bp_systolic'] as num?)?.toInt(),
        bpDiastolic: (json['bp_diastolic'] as num?)?.toInt(),
        pulseBpm: (json['pulse_bpm'] as num?)?.toInt(),
        tempC: (json['temp_c'] as num?)?.toDouble(),
        spo2Pct: (json['spo2_pct'] as num?)?.toInt(),
        weightKg: (json['weight_kg'] as num?)?.toDouble(),
        heightCm: (json['height_cm'] as num?)?.toDouble(),
        recordedAt: (json['recorded_at'] ?? '') as String,
      );

  final int encounterId;
  final int? bpSystolic;
  final int? bpDiastolic;
  final int? pulseBpm;
  final double? tempC;
  final int? spo2Pct;
  final double? weightKg;
  final double? heightCm;
  final String recordedAt;
}

class PreviousHeightWeight {
  PreviousHeightWeight({
    required this.sourceEncounterId,
    required this.recordedAt,
    this.weightKg,
    this.heightCm,
  });

  factory PreviousHeightWeight.fromJson(Map<String, dynamic> json) =>
      PreviousHeightWeight(
        sourceEncounterId: (json['source_encounter_id'] ?? 0) as int,
        weightKg: (json['weight_kg'] as num?)?.toDouble(),
        heightCm: (json['height_cm'] as num?)?.toDouble(),
        recordedAt: (json['recorded_at'] ?? '') as String,
      );

  final int sourceEncounterId;
  final double? weightKg;
  final double? heightCm;
  final String recordedAt;
}

class ClinicTreatment {
  ClinicTreatment({
    required this.id,
    required this.encounterId,
    required this.type,
    required this.description,
    required this.administeredAt,
    this.medicineId,
    this.medicineName,
    this.unit,
    this.quantityUsed,
  });

  factory ClinicTreatment.fromJson(Map<String, dynamic> json) =>
      ClinicTreatment(
        id: (json['id'] ?? 0) as int,
        encounterId: (json['encounter_id'] ?? 0) as int,
        type: (json['treatment_type'] ?? '') as String,
        description: (json['description'] ?? '') as String,
        medicineId: json['medicine_id'] as int?,
        medicineName: json['medicine_name'] as String?,
        unit: json['unit'] as String?,
        quantityUsed: (json['quantity_used'] as num?)?.toInt(),
        administeredAt: (json['administered_at'] ?? '') as String,
      );

  final int id;
  final int encounterId;
  final String type;
  final String description;
  final int? medicineId;
  final String? medicineName;
  final String? unit;
  final int? quantityUsed;
  final String administeredAt;
}
